In [14]:
%pip install python-dotenv langchain_core langchain_google_genai


  Using cached googleapis_common_protos-1.66.0-py2.py3-none-any.whl.metadata (1.5 kB)
  Using cached pyasn1_modules-0.4.1-py3-none-any.whl.metadata (3.5 kB)
  Using cached rsa-4.9-py3-none-any.whl.metadata (4.2 kB)
  Using cached httplib2-0.22.0-py3-none-any.whl.metadata (2.6 kB)
  Using cached google_auth_httplib2-0.2.0-py2.py3-none-any.whl.metadata (2.2 kB)
  Using cached uritemplate-4.1.1-py2.py3-none-any.whl.metadata (2.9 kB)
  Using cached pyasn1-0.6.1-py3-none-any.whl.metadata (8.4 kB)
   ---------------------------------------- 0.0/1.3 MB ? eta -:--:--
   ---------------------------------------- 1.3/1.3 MB 33.3 MB/s eta 0:00:00
   ---------------------------------------- 0.0/12.8 MB ? eta -:--:--
   ------------------------ --------------- 7.9/12.8 MB 40.7 MB/s eta 0:00:01
   -------------------------------- ------- 10.5/12.8 MB 27.3 MB/s eta 0:00:01
   ------------------------------------- -- 12.1/12.8 MB 19.9 MB/s eta 0:00:01
   ---------------------------------------- 12.8/1

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
onnxconverter-common 1.14.0 requires protobuf==3.20.2, but you have protobuf 5.29.3 which is incompatible.
tensorflow-intel 2.17.0 requires protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.20.3, but you have protobuf 5.29.3 which is incompatible.

[notice] A new release of pip is available: 24.3.1 -> 25.0
[notice] To update, run: python.exe -m pip install --upgrade pip


In [15]:
from dotenv import load_dotenv
import os

In [16]:
env_path = '../.env'
load_dotenv(dotenv_path=env_path)

True

In [20]:
GOOGLEAI_API_KEY = os.getenv("GOOGLE_API_KEY")

In [21]:
from langchain_google_genai import ChatGoogleGenerativeAI

llm = ChatGoogleGenerativeAI(model="gemini-2.0-flash")

In [22]:
llm.invoke("What is the meaning of life?")

AIMessage(content='Ah, the age-old question! There\'s no single, universally accepted answer to the meaning of life. It\'s a deeply personal and philosophical question that has been pondered by thinkers for centuries. Here\'s a breakdown of some common perspectives:\n\n**1. Nihilism:**\n\n*   **Core Idea:** Life is inherently without objective meaning, purpose, or intrinsic value.\n*   **Implication:** We are free to create our own meaning, but ultimately, nothing truly matters in the grand scheme of the universe.\n\n**2. Existentialism:**\n\n*   **Core Idea:** Existence precedes essence. We are born into the world without a predetermined purpose, and it is up to us to define our own meaning through our choices and actions.\n*   **Implication:** We have freedom and responsibility to create our own values and live authentically.\n\n**3. Absurdism:**\n\n*   **Core Idea:** The search for meaning in a meaningless universe is inherently absurd. There\'s a fundamental conflict between our de

Prompt templating:

In [30]:
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.output_parsers import StrOutputParser

def_prompt = ChatPromptTemplate.from_messages([
    ("system", "You are a world class defense attorney."),
    ("human", "Can you tell me {topic}")
])

In [31]:
def_prompt.invoke({"topic": "Can you tell me a good defense strategy for someone accused of stealing cookies?"})

ChatPromptValue(messages=[SystemMessage(content='You are a world class defense attorney.', additional_kwargs={}, response_metadata={}), HumanMessage(content='Can you tell me Can you tell me a good defense strategy for someone accused of stealing cookies?', additional_kwargs={}, response_metadata={})])

Chaining via pipe:

In [32]:
chain = def_prompt | llm | StrOutputParser()

In [33]:
chain.invoke({"topic": "Can you tell me a good defense strategy for someone accused of stealing cookies?"})

'Alright, let\'s craft a defense strategy for someone accused of cookie theft. The best approach will depend on the specific facts of the case, but here are a few potential avenues we can explore, ranging from the lighthearted to the more serious:\n\n**I. The "Mistaken Identity/Circumstantial Evidence" Defense:**\n\n*   **Focus:** This is often the most practical and widely applicable. The goal is to create reasonable doubt that your client was the actual cookie thief.\n*   **Key Arguments:**\n    *   **Lack of Direct Evidence:** Emphasize that there is no direct evidence (video, eyewitness testimony, etc.) definitively placing your client at the scene of the crime *with* the cookies. Any evidence is circumstantial.\n    *   **Mistaken Identity:**  Argue that someone else could have easily been mistaken for your client.  Do they have a similar build, clothing style, or mannerisms as someone else who was in the vicinity?  "My client may have been *near* the cookie jar, but that doesn\'t

Streaming chunks:

In [36]:
for chunk in chain.stream({"topic": "How would you prove that someone didn’t steal the cookies when circumstantial evidence suggests otherwise?"}):
    print(chunk, end="|")

Alright|, let's assume my client is accused of stealing cookies and the prosecution has| a case built on circumstantial evidence. My job is to create reasonable doubt in the| minds of the jury. Here's how I'd approach it:

**1. Dismantling the Circumstantial Evidence, Piece by Piece:**|*   **Focus on Alternative Explanations:** The core strategy is to offer believable alternative explanations for *every* piece of circumstantial evidence.  We need to show| that the evidence, while suggestive, doesn't *necessarily* point to my client.

    *   **Example:** Let's say the circumstantial evidence includes:
        *   **Cookie crumbs on my client's shirt:**  "|Ladies and gentlemen, my client works in a bakery! Cookie crumbs are an occupational hazard.  We will present testimony from coworkers who will attest to the ubiquitous nature of crumbs in that environment.  Furthermore, the prosecution hasn't presented any| evidence linking *these specific* crumbs to *those specific* cookies. They co

Alright|, let's assume my client is accused of stealing cookies and the prosecution has| a case built on circumstantial evidence. My job is to create reasonable doubt in the| minds of the jury. Here's how I'd approach it:

**1. Dismantling the Circumstantial Evidence, Piece by Piece:**|*   **Focus on Alternative Explanations:** The core strategy is to offer believable alternative explanations for *every* piece of circumstantial evidence.  We need to show| that the evidence, while suggestive, doesn't *necessarily* point to my client.

    *   **Example:** Let's say the circumstantial evidence includes:
        *   **Cookie crumbs on my client's shirt:**  "|Ladies and gentlemen, my client works in a bakery! Cookie crumbs are an occupational hazard.  We will present testimony from coworkers who will attest to the ubiquitous nature of crumbs in that environment.  Furthermore, the prosecution hasn't presented any| evidence linking *these specific* crumbs to *those specific* cookies. They could be from *any* cookie."
        *   **My client was seen near the cookie jar:** "Being *near* a cookie jar is not a crime. My client was in the kitchen to get a glass of water.  |We will present evidence that the water cooler is located directly next to the cookie jar.  The prosecution is asking you to jump to conclusions based on mere proximity."
        *   **The cookie jar was empty shortly after my client was in the kitchen:** "Correlation does not equal causation. Many people were in the kitchen that| day.  The prosecution hasn't ruled out *any* other potential cookie thief. Perhaps someone else took the cookies earlier and my client simply walked in to an empty jar.  The timing is coincidental, not proof."
        *   **My client has a known sweet tooth:** "Having a sweet tooth is a common| human trait!  It doesn't make someone a thief.  The prosecution is attempting to prejudice the jury by painting my client as someone who *might* steal cookies, but 'might' is not 'did'."

*   **Highlighting Weaknesses in the Prosecution's Evidence:**  I'd meticulously| examine the prosecution's evidence for flaws and inconsistencies.

    *   **Chain of Custody:** "How were these cookie crumbs collected?  Who handled them?  Can the prosecution definitively prove that the crumbs presented in court are the same crumbs found on my client's shirt, and that they weren't contaminated| in any way?  Gaps in the chain of custody create doubt."
    *   **Witness Credibility:** "Did the witness have a clear view? Were they biased in any way?  Did they have any reason to lie or exaggerate?  We will explore any potential motives for the witness to misremember| or misrepresent the events."
    *   **Lack of Direct Evidence:** "The prosecution has presented *no* direct evidence. No one saw my client take the cookies. There are no fingerprints on the jar. There is no confession.  This case rests entirely on assumptions and speculation."

**2. Establishing| an Alibi (If Possible):**

*   **Solid Alibi:** If my client has a verifiable alibi – meaning they were demonstrably somewhere else when the cookies were taken – this is powerful.  We'd present witnesses, receipts, security footage, or any other evidence to support the alibi.|
*   **Weak Alibi Still Helps:** Even if the alibi isn't airtight, it can still create doubt.  "While my client cannot definitively prove they were elsewhere, they were engaged in [activity] at the approximate time of the cookie theft.  This makes it less likely that they had the| opportunity to commit the alleged act."

**3. Presenting Affirmative Evidence of Innocence:**

*   **Good Character:**  "My client is a person of good character.  We will present witnesses who will testify to their honesty, integrity, and lack of any prior history of theft or dishonesty.  This| is not the type of person who would steal cookies."
*   **Alternative Suspects:**  Without directly accusing anyone, I might subtly introduce the possibility of other potential cookie thieves.  "The prosecution has focused solely on my client, but were there other individuals who had access to the cookies? Did the prosecution| thoroughly investigate all possible leads?"
...

*   **Tailor the Defense to the Specific Facts:**  The best defense is always tailored to the unique circumstances of the case.  I'd need to thoroughly investigate the facts and interview witnesses to develop the most effective strategy.
*   |**Ethical Obligations:**  I cannot knowingly present false evidence or encourage perjury. My role is to vigorously defend my client within the bounds of the law and ethical rules.

**In conclusion,** proving a negative ("I didn't do it") is always challenging. My approach would be to systematically undermine the prosecution's| circumstantial evidence, present alternative explanations, highlight the weaknesses in their case, and create reasonable doubt in the minds of the jury. The goal is to show that the evidence, while suggestive, does not definitively prove that my client stole the cookies.|

Context-aware generation:

In [37]:
from datetime import date
CAG_prompt = ChatPromptTemplate.from_messages([
    ("system", 'You know that the current date is {date}.'),
    ("human", "{question}")
])

In [42]:
CAG_chain = CAG_prompt | llm | StrOutputParser()
CAG_chain.invoke({"question": "How many days have passed since New Year's day till today?", "date": date.today()})

'From January 1st, 2025 to February 7th, 2025, there are 31 days in January and 7 days in February. So the total number of days is 31 + 7 = 38 days.'

In [54]:
SOURCE_RAG_EXP1 = """
Phoenix Inc. 2024 R&D initiatives, performance metrics, and financial details:

- Q1: 
  - Developed an AI-powered analytics tool that enhanced client decision-making capabilities, leading to a 20% reduction in operational costs for clients.
  - Conducted 50+ research papers on emerging machine learning algorithms.
  - Budget: $125,000, Revenue Generated: $300,000
  - Key Metrics: 15 new patents filed, 100 client demos conducted, 5 collaborations initiated.

- Q2:
  - Released the beta version of a cloud-based collaboration platform, which garnered 10,000 early users, of which 20% converted to premium subscriptions.
  - Developed machine learning models for demand forecasting in supply chains, improving forecast accuracy by 35%.
  - Budget: $210,000, Revenue Generated: $450,000
  - Key Metrics: 50% user satisfaction, 200+ features requested by beta testers, 2 new strategic partnerships.

- Q3:
  - Overhauled cybersecurity systems with AI-driven monitoring, preventing 15 major cyberattacks, and resulting in zero security breaches.
  - Launched an industry-first AI tool for predictive maintenance, reducing equipment downtime by 25% across all client facilities.
  - Budget: $180,000, Revenue Generated: $500,000
  - Key Metrics: 10% reduction in client operational costs, 30% reduction in downtime, 3 industry awards won.

- Q4:
  - Conducted large-scale user testing and feedback analysis, leading to 85 actionable insights that shaped the final product versions.
  - Initiated a cross-department collaboration initiative to integrate AI insights into all business units.
  - Budget: $95,000, Revenue Generated: $450,000
  - Key Metrics: 90% client retention rate, 10% improvement in overall product usability, 5 new industry certifications earned.

Financial Overview:
- Total R&D Budget for 2023: $610,000
- Total Revenue Generated from R&D Products: $1.7 Million
- Gross Margin from R&D Projects: 45%
- Estimated Total Cost Savings for Clients from R&D Initiatives: $2.5 Million
- Projected Revenue Growth in 2024 from R&D Innovations: 18%

Strategic Overview:
- The AI-powered tools developed in 2023 are expected to reduce client operational costs by 15% across all sectors.
- Long-term forecast suggests a 20% improvement in overall efficiency for clients utilizing Phoenix's R&D products by 2025.
"""

In [55]:
rag_prompt = ChatPromptTemplate.from_messages([
    ("system", 
     "You are a highly skilled AI assistant specialized in advanced business analysis, forecasting, and financial modeling. Utilize the provided context to generate comprehensive, data-driven insights, perform complex calculations, and offer strategic recommendations based on the available information:\n\n{context}"),
    ("human", 
     "Considering the data provided, please answer the following multi-faceted question with detailed analysis and reasoning:\n\n{question}")
])

In [56]:
rag_chain = rag_prompt | llm | StrOutputParser()

In [57]:
result = rag_chain.invoke({"context": SOURCE_RAG_EXP1, "question": "Considering Phoenix's 2023 R&D performance and financial metrics, calculate the ROI for each quarter, identify the quarter with the highest ROI, and explain the factors contributing to the most successful quarter. Additionally, evaluate the long-term strategic impact of the AI-powered tools developed, including projected efficiency gains and revenue growth. How would these metrics shape Phoenix's strategic decisions for 2024?"})

In [58]:
print(result)

Okay, let's break down Phoenix Inc.'s 2023 R&D performance and formulate strategic recommendations for 2024.

**1. ROI Calculation and Analysis by Quarter:**

ROI (Return on Investment) is calculated as:  `ROI = ((Revenue Generated - Budget) / Budget) * 100`

*   **Q1:**
    *   Budget: $125,000
    *   Revenue: $300,000
    *   ROI: (($300,000 - $125,000) / $125,000) * 100 = **140%**

*   **Q2:**
    *   Budget: $210,000
    *   Revenue: $450,000
    *   ROI: (($450,000 - $210,000) / $210,000) * 100 = **114.29%**

*   **Q3:**
    *   Budget: $180,000
    *   Revenue: $500,000
    *   ROI: (($500,000 - $180,000) / $180,000) * 100 = **177.78%**

*   **Q4:**
    *   Budget: $95,000
    *   Revenue: $450,000
    *   ROI: (($450,000 - $95,000) / $95,000) * 100 = **373.68%**

**Conclusion:** Q4 has the highest ROI at 373.68%.

**2. Factors Contributing to the Most Successful Quarter (Q4):**

Q4's exceptional ROI can be attributed to several factors:

*   **Lower Budget:** Q4 had the lowest 

Okay, let's break down Phoenix Inc.'s 2023 R&D performance and formulate strategic recommendations for 2024.

**1. ROI Calculation and Analysis by Quarter:**

ROI (Return on Investment) is calculated as:  `ROI = ((Revenue Generated - Budget) / Budget) * 100`

*   **Q1:**
    *   Budget: $125,000
    *   Revenue: $300,000
    *   ROI: (($300,000 - $125,000) / $125,000) * 100 = **140%**

*   **Q2:**
    *   Budget: $210,000
    *   Revenue: $450,000
    *   ROI: (($450,000 - $210,000) / $210,000) * 100 = **114.29%**

*   **Q3:**
    *   Budget: $180,000
    *   Revenue: $500,000
    *   ROI: (($500,000 - $180,000) / $180,000) * 100 = **177.78%**

*   **Q4:**
    *   Budget: $95,000
    *   Revenue: $450,000
    *   ROI: (($450,000 - $95,000) / $95,000) * 100 = **373.68%**
...
*   **Capitalize on Industry Recognition:** The three industry awards won in Q3 should be leveraged for marketing and public relations efforts to enhance brand reputation and attract new clients. Highlight the success stories of clients who have benefited from Phoenix's AI-powered solutions.
*   **Data-Driven Decision Making:** Continue the trend of using AI-powered analytics tools to make informed decisions, both internally and for clients. This demonstrates the value of Phoenix's core competency and drives competitive advantage.

**In summary, Phoenix Inc. should strategically focus on maximizing the impact of its existing AI-powered tools in 2024 through refinement, integration, and strong client relationships. By allocating resources strategically and leveraging key metrics, Phoenix can sustain revenue growth, improve client efficiency, and solidify its position as a leader in AI-driven solutions.**